### DISCUSSION TITLE
help file progress

### DISCUSSION THREAD

**Discussion by davidgritter** [GrandOrgue/GoOdf#4] [General]
I've created a first draft with most of the text needed to walk people through the process of creating an odf file.  I still need to do the section on the panel creation.  Then what remains is a page for every one of the objects used.  Hopefully I can send you a draft late this week or early next week.

**Comment by larspalo** [comment]
Excellent! Looking forward to it!

**Comment by davidgritter** [comment]
I have several questions that have come up in writing the help file
If I create a manual and then create a GUI element from the manual, I see at the bottom of  the window when editing the GUI element
DisplayKeY||maps to backend MIDI||and front end MIDI.   Can you explain?

I often end  up losing data when entering numbers in  a box.  This appears to be because the data doesn't get accepted until you tab out or click outside the data entry box, but not inside an action button.  my expectation would be that data is accepted whenever a box loses focus, such as switching to another object or hitting an exit button.  Is this a WXwidget limitation, and if so what is the best way to describe in the help a way to insure that data is not lost?

I do not have a good explaination for the parameters at the bottom of the manual editing window:
Midi key map key|| maps to value||
Can you explain?
Thanks

**Reply by larspalo** [reply]
For the structural (backend) manual section you have:

> MIDIKey000 - MIDIKey127
> (integer 0 - 127, default: same MIDI key number) Allows to map the MIDI note in MIDIKey999 to a different number. This mapping is used by the default manual MIDI matching type - others may or may not use this mapping table.  

For the frontend gui manual you have:

> DisplayFirstNote
> (integer 0 - 127, default: FirstAccessibleKeyMIDINoteNumber) Display the first key as the following note. 
> DisplayKeys
> (integer 1 - NumberOfAccessibleKeys, default: NumberOfAccessibleKeys) number of keys to display. 
> DisplayKey999
> (integer 0 - 127, default: FirstAccessibleKeyMIDINoteNumber + 999) The number in the key (999) is between 1 and DisplayKeys. It contains the MIDI number of the backend key, that is connected to this GUI key. 
> DisplayKey999Note
> (integer 0 - 127, default: FirstAccessibleKeyMIDINoteNumber + 999) The number in the key (999) is between 1 and DisplayKeys. It contains the MIDI number of the displayed frontend GUI key. 

This allows for quite sophisticated mapping of displayed keys to backend MIDI keys that actually triggers the correct pipes. Normally, **for most organs, one should just leave the default mapping values as they are** and never bother with them. But for special type keyboards, like in the Norrfjärden Sample Set I've done, you can have different keyboard layouts (broken octave in the manuals, and shortened pedal keyboard layout) that will play different notes on some keys than what's apparent for anyone that's only accustomed to modern type keyboards. That is one use case for needing this option.

The only way to be sure that data is entered and applied (especially in any kind of text box that have numbers), one need to leave it or otherwise confirm the change of content. In a few cases I've made sure to catch any change happening to process the input values, but I think that in the documentation we should just add a comment that one should always remember to apply the manually changed values (that then can be validated by the program) by focusing out of the control somehow. While it's possible to program the controls to catch everything that happens in a control and continually make the validations/applying of values, it kind of makes more sense to allow the user to finish writing the values and then do the checking/applying.

**Comment by davidgritter** [comment]
Attached is the first cut at a help file.  The heading in red at the top of the file attempts to show the html file hierarchy intended.  I am able to export to html from this file where every Header1 starts a new html file.  These individual html files will need some work to redefine hyperlinks, but I have been able to see them in GOODF help window.
[GOODFhelp.odt](https://github.com/larspalo/GOODF/files/10963669/GOODFhelp.odt)

**Reply by larspalo** [reply]
I've browsed through the help you've written, and apart from a few small typos I think it's very good indeed. If you have (or make) a clone of the GOODF repository you can create a branch for the help, export the text into separate html files, commit to the help branch and create a PR to GOODF repo from it. The hyperlinks can be edited when you have separated the text into different files. Just ask if you need any help with it.

**Comment by davidgritter** [comment]
I have a working help system now, with contents and index and search all working.  I have a couple of questions, though.   I cannot get text to wrap around images in help, even thought it wraps properly when viewed in a web browser.  The resolution/clarity of the GOODF screen shot images is not very good in help, even though they are clearly visible in a web browser.  Are there parameters to adjust this in the help system, or do I need to make my images bigger and forget about wrapping text?

I cannot make a new branch for your project. Is it OK if I just send a zip file of what needs to go in the help directory?

**Reply by larspalo** [reply]
The wxHTML is somewhat limited, especially when it comes to styling (see https://docs.wxwidgets.org/3.0/overview_html.html#overview_html_helpformats). Keep the image files of screenshots as original size. Text wrapping should be possible to arrange, but it might be better to just show the images as blocks. You can experiment with things like:

`<p><img style="width: 100%;" src="images/TheImage.jpg" alt="An image of something special"></p>`

or test different align options and percentages. In the worst case you could even resort to using a `<table>` like in the old days to control the layout...

But I think that generally it's best to keep things as simple as possible and not bother too much with the styling. After all, this is a utility program, not a contender for the most beautiful GUI out there! Most important is that there will be a clear help that covers the important things in the program that the user might not immediately understand.

If you have trouble with git and github (I'm still very much learning it myself) I can of course make the commit once you're happy with the content. Thanks for your help!

**Comment by davidgritter** [comment]
attached is a zip of the help directory for GOODF.  It replaces the existing help directory in the source tree, and should result in a working help system.  The trick to getting the graphics right, as you suggested, was to use them full size.
[help.zip](https://github.com/larspalo/GOODF/files/11005945/help.zip)

**Reply by larspalo** [reply]
I've committed the help files. I moved the images to a separate directory and updated the img tags src value to correspond. Thanks for your work with the help! This in itself would merit a new release but since I'm fairly close to having the .organ file reading completed I think I'll wait until that's working too. Also I'm going to look into the possiblity of automating the release builds here on GitHub instead of doing them manually.

**Reply by larspalo** [reply]
@davidgritter New builds are now provided for the releases by GitHub runners. Since you can build GOODF yourself, you're not really affected by this, but your help files are now included in the v0.4.0 release.

**Reply by Jean-Gabriel1982** [reply]
Hi,
Is there a way to have the Help on pdf ? I want to translate it in French ..

I'll try to make an organ from scratch 😀 !

Thank you !

**Comment by davidgritter** [comment]
I've attached an old help version in pdf (original help was written in libreoffice), but it is probably out of date.  You can also download the source files for GOODF, change to the  help directory, then open each .html file in a browser or in libreoffice, then print or export to pdf.
[GOODFhelp.pdf](https://github.com/larspalo/GOODF/files/11772818/GOODFhelp.pdf)

**Reply by larspalo** [reply]
@Jean-Gabriel1982 I've made some updates to the help files since @davidgritter created them. The best way to translate them at the moment is to do as David suggests and open the .html help files (in the GOODF sources there's a help directory containing all the files) in a web browser and then copy/paste it into Google translate for a rough start.

Using GOODF is fairly simple though, there are a few shortcuts and procedures that might not be so obvious to get immediately, but I think most of them are mentioned in the help. Eventually I'll create a video screencast series about how to use GOODF, but if you have any questions - just ask!

**Reply by Jean-Gabriel1982** [reply]
@davidgritter @larspalo 
Many thanks for your answer. I think I'll do as you suggest. 

When I've opened the program for the first time, I was a bit afraid of all fields we add to fill 😄 
But I think that is not the program but my understanding of organ building that fails 😁
